### PR TITLE
Bugfix: Fix parsing array values from osx_defaults

### PR DESCRIPTION
**Pull request by notok**
Unquote values and unescape double quotes when reading array values from osx defaults.

##### SUMMARY

When reading array values from defaults on osx, some value may be double-quoted.
But osx_defaults module do not unquote them currently.
In some cases, array_add option may not idempotent because of this.

Additionally, double quotes in the value is escaped like `\\"`.
They also should be unescaped.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

osx_defaults

##### ADDITIONAL INFORMATION

For example, the task below adds value every time even if the value is already added.

```
- osx_defaults:
    domain: com.apple.systemuiserver
    key: menuExtras
    value:
      - '/System/Library/CoreServices/Menu Extras/Volume.menu'
    type: array
    array_add: yes
```

Values read from osx is double-quoted, and it's different from the value in task(not quoted).
So ansible always try to add the value.

### CODE REVIEW

**Comment by notok** [comment]
I tried to keep the code shorter, but it may lacks readability.
Any advices are welcome. Thanks.

**Comment by felixfontein** [comment]
This needs a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to).

**Comment by notok** [comment]
@felixfontein Thanks for your comment.
Added changelog fragment. Please recheck.

File: changelogs/fragments/fix_parsing_array_values_in_osx_defaults.yml
Comment (felixfontein) [inline changelogs/fragments/fix_parsing_array_values_in_osx_defaults.yml:2]:
```suggestion
  - osx_defaults - unquote values and unescape double quotes when reading array values (https://github.com/ansible-collections/community.general/pull/358).
```

**Comment by felixfontein** [comment]
Changelog fragment looks good. Cannot comment on the rest. Maintainers, please take a look.

**Comment by gundalow** [comment]
@MorrisA @Akasurde @bcoca @d-little @flynn1973 @gforster @kairoaraujo @kyleabenson @martinm82 @marvin-sinister @mator @molekuul @ramooncamacho @wtcross Hi, I was wondering if you'd be able to take a look at this small bugfix for `osx_say`. Thanks in advance.

**Comment by gundalow** [comment]
recheck

**Comment by notok** [comment]
I'm adding more detailed description for this problem. A little bit long..

The macos defaults command returns list value with following format.

```
$ defaults write testdomain testkey -array testvalue1 testvalue2
$ defaults read testdomain testkey
(
    testvalue1,
    testvalue2
)
```

And, if the value in the list includes non-alphanumeric string like underscore `_`, or whitespace ` `, they are double-quoted.
Moreover, if the value includes double-quotes, the double-quotes are escaped like `\\"`.

```
$ defaults write testdomain testkey -array non_alpnanumeric "white space" -string 'value with "double-quotes"'
$ defaults read testdomain testkey
(
    "non_alpnanumeric",
    "white space",
    "value with \\"double-quotes\\""
)
```

Currently, this `osx_defaults` module does NOT unquote or unescape the value when parsing list values.

This makes ansible to unexpectedly detect changes.

e.g. The following task compares specified value `/System/Library/CoreServices/Menu Extras/Volume.menu` with value from defaults `"/System/Library/CoreServices/Menu Extras/Volume.menu"`, and detected as changed. And always adds the value even if already exists in the list.

```
- osx_defaults:
    domain: com.apple.systemuiserver
    key: menuExtras
    value:
      - '/System/Library/CoreServices/Menu Extras/Volume.menu'
    type: array
    array_add: yes
```

**Comment by notok** [comment]
@MorrisA @Akasurde @bcoca @d-little @flynn1973 @gforster @kairoaraujo @kyleabenson @martinm82 @marvin-sinister @mator @molekuul @ramooncamacho @wtcross

Hello. Could you please check this small bugfix for osx_defaults?
This is just a one-line fix, and I've written a detailed description on the issue comment.
Thanks in advance.

**Comment by Andersson007** [comment]
@frenck could you please take a look?

**Comment by Andersson007** [comment]
@notok hi, thanks for the fix! Could you please email the author? there's the email in the module's header.

**Comment by frenck** [comment]
I don't think so. Generally, I don't appreciate the random tag. Thanks.

**Comment by Andersson007** [comment]
@frenck  thanks for the reply. i'm not sure i understand if you're for or against these changes?

**Comment by felixfontein** [comment]
@Andersson007 I think frenck does not want to be contacted about this module.

**Comment by frenck** [comment]
@Andersson007 It means I have no opinion about this matter. I don't use Ansible anymore. Haven't contributed in years and don't plan on doing so. Yet, I keep being pulled in by random people.

Pulling in people at random to a conversation is IMHO impolite. For example, I do around 50-100 PR reviews a day. I'm happy to ping you in all of them?

**Comment by Andersson007** [comment]
@frenck if those 50-100 PRs relate to stuff i created, no problem.
Ok, let's wait anybody else to review

**Comment by frenck** [comment]
I'm no supporting this module or project anymore. I'm happy to put in a PR to remove this module, if that is needed to stop this madness.

I've blocked the Ansible org and the bullbot for the same reason, as it seems impossible to get rid of this project.

**Comment by felixfontein** [comment]
@frenck I've created a PR to remove your github handle from the module and BOTMETA, and add a mark to your email address in the module's copyright header stating that you do not wished to be contacted. I hope this will be the last time you will be pinged because of this module.

**Comment by Andersson007** [comment]
to push Notifications "Unsubscribe" on the right vertical bar should also work to prevent getting anything from this PR

**Comment by frenck** [comment]
I did until I was mentioned again 😉

File: plugins/modules/system/osx_defaults.py
Comment (felixfontein) [inline plugins/modules/system/osx_defaults.py:233]: Can extra spaces at the end (after the comma) not appear? Otherwise,
```suggestion
        value = [re.sub('^ *"?|"?,? *$', '', x.replace('\\"', '"')) for x in value]
```
might be better. (The original code also handled spaces after the comma.)

**Comment by martinmigasiewicz-tomtom** [comment]
@notok could you add as well an integration test covering this case in https://github.com/ansible-collections/community.general/blob/main/tests/integration/targets/osx_defaults/tasks/main.yml.

Having the test in place will (hopefully) make the regex change a bit clearer

**Comment by martinm82** [comment]
@notok could you add as well an integration test covering this case in https://github.com/ansible-collections/community.general/blob/main/tests/integration/targets/osx_defaults/tasks/main.yml.

Having the test in place will (hopefully) make the regex change a bit clearer

**Comment by dmsimard** [comment]
Hi @notok, this looks almost ready to merge if the latest comments are addressed -- let us know if we can help. Thanks !

**Review by felixfontein — APPROVED** [review]

**Review by Andersson007 — APPROVED** [review]

**Comment by Andersson007** [comment]
@notok after this PR merged, would you like to be added as a maintainer of the module to the `.github/BOTMETA.yml` file?
Your shipit/LGTM will be counted by the bot for automerge of fixes/new features and you'll get mail notification about issues/PRs related to the module (i don't think there will be to many mails and nothing obligatory, of course). What do you think?

**Comment by Andersson007** [comment]
@martinm82 @dmsimard do you have anything to suggest or we can merge?

**Comment by dmsimard** [comment]
LGTM but I defer to @martinm82 :)

**Review by martinm82 — APPROVED** [review]
LGTM

**Comment by Andersson007** [comment]
@notok thanks for the contribution! and I'm still waiting for you feedback about maintainership:)
@felixfontein @gundalow @dmsimard @martinm82 thanks for reviewing!

**Comment by notok** [comment]
@Andersson007 Thanks for inviting. Could you add me to BOTMETA.yml?
I'd like to help if there's a chance (and I also think not so many issues/PRs will come).

Thanks all for reviewing!

**Pull request merged**
